pkg/solana/chainreader/bindings.go

```diff
@@ -13,167 +14,420 @@ import (
+	if binding == nil {
 		return fmt.Errorf("%w: bound contract is nil", types.ErrInvalidType)
 	}
 
-	if err := b.handleAddressSharing(boundContract); err != nil {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+
+	if err := r.handleAddressSharing(binding); err != nil {
 		return err
 	}
 
-	rBindings, nameSpaceExists := b.namespaceBindings[boundContract.Name]
-	if !nameSpaceExists {
-		return fmt.Errorf("%w: no namespace named: %q", types.ErrInvalidConfig, boundContract.Name)
+	namespace, nbsExist := r.namespaceBindings[binding.Name]
+	if !nbsExist {
+		return fmt.Errorf("%w: no namespace named %s", types.ErrInvalidConfig, binding.Name)
 	}
 
-	key, err := solana.PublicKeyFromBase58(boundContract.Address)
+	address, err := solana.PublicKeyFromBase58(binding.Address)
 	if err != nil {
-		return fmt.Errorf("%w: failed to parse address: %q for contract %q", types.ErrInvalidConfig, boundContract.Address, boundContract.Name)
+		return err
 	}
 
-	for _, rBinding := range rBindings {
-		rBinding.SetAddress(key)
+	return errors.Join(
+		namespace.Bind(ctx, reg, address),
+		namespace.BindReaders(ctx, address),
+	)
+}
+
+func (r *bindingsRegistry) Unbind(ctx context.Context, reg filterRegistrar, binding types.BoundContract) error {
+	r.mu.RLock()
+	defer r.mu.RUnlock()
+
+	namespace, nbsExist := r.namespaceBindings[binding.Name]
+	if !nbsExist {
+		return fmt.Errorf("%w: no namespace named %s", types.ErrInvalidConfig, binding.Name)
 	}
 
-	return nil
+	return errors.Join(
+		namespace.Unbind(ctx, reg),
+		namespace.UnbindReaders(ctx),
+	)
 }
 
-func (b *bindingsRegistry) SetCodecs(codec types.RemoteCodec) {
-	for _, nbs := range b.namespaceBindings {
-		for _, rb := range nbs {
-			rb.SetCodec(codec)
-		}
+func (r *bindingsRegistry) CreateType(namespace, readName string, forEncoding bool) (any, error) {
+	rBinding, err := r.GetReader(namespace, readName)
+	if err != nil {
+		return nil, err
 	}
+
+	return rBinding.CreateType(forEncoding)
 }
 
-func (b *bindingsRegistry) SetModifiers(modifier commoncodec.Modifier) {
-	for _, nbs := range b.namespaceBindings {
-		for _, rb := range nbs {
-			rb.SetModifier(modifier)
+func (r *bindingsRegistry) initAddressSharing(addressShareGroups [][]string) error {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+
+	r.addressShareGroups = make(map[string]*addressShareGroup)
+
+	for _, group := range addressShareGroups {
+		shareGroup := &addressShareGroup{
+			address: solana.PublicKey{},
+			group:   group,
+		}
+
+		for _, namespace := range group {
+			if _, alreadySharesAddress := r.addressShareGroups[namespace]; alreadySharesAddress {
+				return fmt.Errorf("namespace %q can't share address with two different groups", namespace)
+			}
+
+			r.addressShareGroups[namespace] = shareGroup
 		}
 	}
+
+	return nil
+}
+
+func (r *bindingsRegistry) getShareGroup(nameSpace string) (*addressShareGroup, bool) {
+	shareGroup, sharesAddress := r.addressShareGroups[nameSpace]
```

Comment:
Wouldn't we need a read lock here?

Reply:
It would be safe, but the groups don't change after initialization.

Reply:
fixed